docs/source/plugin/internals.rst

```diff
@@ -54,3 +54,16 @@ sopel.plugins.rules
    .. autoclass:: AbstractNamedRule
       :members:
       :undoc-members:
+
+   .. class:: TypedRule
+
+         A :class:`~typing.TypeVar` bound to :class:`AbstractRule`. When used
+         in the :meth:`AbstractRule.from_callable` class method, it means the
+         return value must be an instance of the class used to call that method
+         and not a different subclass of ``AbstractRule``.
+
+         .. versionadded:: 8.0
+
+            This ``TypeVar`` was added in an effort to add type-hint to Sopel.
```

Comment:
```suggestion
             This ``TypeVar`` was added as part of a goal to start type-checking Sopel.
```

Maybe even go as far as saying that it's not used at runtime? I don't know how many plugin devs look here 😂